docs/rfcs/component-status-reporting.md

```diff
@@ -0,0 +1,196 @@
+5. Component health reporting must only be a mechanism for reporting health - it should have no mechanisms for taking actions on the health it reports.  How consumers of the health reporting system respond to component updates is not a concern of the health reporting system.
+
+## Existing deviations from those goals
+
+### Fatal Error Reporting
+
+Before the current implementation of component status reporting, a component could stop the collector by using `Host.ReportFatalError`. Now, a component MUST use component status reporting and emit a `FatalError`. This fact is in conflict with Goal 1, which states component health reporting must be opt-in for components.
+
+A couple solutions:
+1. Accept this reality as an exception to Goal 1.
+2. Add back `Host.ReportFatalError`.
+3. Remove the ability for components to stop the collector be removing `FatalError`.
+
+### No way to identify components that are not reporting status
+Goal 1 states that consumers of component status reporting must be able to identify components in use that have not opted in to component status reporting. Our current implementation does not have this feature.
+
+### Should component health reporting be an opt-in for `component.Host` implementations?
+
+The current implementation of component status reporting does not add anything to `component.Host` to force a `component.Host` implementation, such as `service.Service`, to be compatible with component status reporting.  Instead, it adds `ReportStatus func(*StatusEvent)` to `component.TelemetrySettings` and things that instantiate components, such as `serivce.Service`, should, but are not required, to pass in a value for `ReportStatus`.
+
+As a result, `component.Host` implementation is not required to engage with the component status reporting system.  This could lead to situations where a user adds a status watcher extension that can do nothing because the `component.Host` is not reporting component status updates.
+
+Is this acceptable? Should we:
+1. Require the `component.Host` implementations be compatible with the component status reporting framework?
+2. Add some sort of configuration/build flag then enforces the `component.Host` implementation be compatible (or not) with component status reporting?
+3. Accept this edge case.
+
+### Component TelemetrySettings Requirements
+
+The current implementation of component status reporting added a new field to `component.TelemetrySettings`, `ReportStatus`.  This field is technically optional, but would be marked as stable with component 1.0. Are we ok with 1 of the following?
+
+1. Including a component status reporting feature, `component.TelemetrySettings.ReportStatus`, in the 1.0 version of `component.TelemetrySettings`?
+2. Marking `component.TelemetrySettings.ReportStatus` as experimentatal via godoc comments in the 1.0 version of `component.TelemetrySettings`?
+
+Or should we refactor `component` somehow to remove `ReportStatus` from `component.TelemetrySettings`?
+
```

Comment:
My gut feeling is that `ReportStatus` should be either part of `component.Host` or an optional interface to be implemented by `component.Host`s

Reply:
I like the idea of an optional interface that `component.Host`s can implement.

Reply:
The larger goal for components to report runtime status, see #9957. One issue with implementing `ReportStatus` on host, is that it's passed into start, and for a component to have access to `ReportStatus` outside of `Start` it needs to save a handle, either to the function, or to the host. Few do this, but most components save a handle to TelemetrySettings, which provides similar mechanisms for the component to report data about itself. It's not impossible for it be implemented on Host, but its an inconvenient location.

Reply:
>  Few do this

I think this is something we can remediate with good examples, documentation and possibly some autogenerated test where we assert common types of status transitions (e.g. failed to create a server).

The only difference I see is that you can't report status during component creation, but I see that as a feature, not a bug (there's going to be no status transition before a component has been initialized).

Reply:
I am more interested in the ergonomics for component authors. I think its natural for the API to report status to be in the same place as I find a meter, or logger for a component. The extra step of explicitly having to save a separate handle to host, or its ReportStatus method to use outside of Start makes the API more cumbersome to use.

Reply:
The advantages I see if we put this in `component.Host` are:

1. Allowing component status to evolve independently from the rest of the core APIs and decoupling this effort from Collector 1.0 (the most important benefit in my mind)
2. Making it impossible for components to misuse `ReportStatus` during creation time
3. Potentially removing the awkwardness in having both a `component.TelemetrySettings` and `servicetelemetry.TelemetrySettings`, and simplifying the implementation of #4970

I feel like these outweigh `component.TelemetrySettings` benefits, but it would be great to have a similar list so we can consider pros/cons

Reply:
On point 3, we will actually introduce the same awkwardness in a different spot. We'll end up with a `servicehost.Host` and a `component.Host`. This transition from host to telemetry settings was made once before during the original implementation. See this commit for how things looked when ReportStatus was on Host: https://github.com/open-telemetry/opentelemetry-collector/pull/8169/commits/b45540e44bd7ad7d114baa983241f7995d72ca5d.

Reply:
Thanks for the context, I think I wasn't super clear on (3). We have several vendors interested in providing their own meter provider, tracer provider, logger... (see #4970) so moving this awkwardness to `component.Host` would potentially simplify adding support for this: all other fields in `component.TelemetrySettings` are governed by the `service::telemetry` config section and we could more easily use a factory pattern if we were left with the other fields. That said, I don't think (3) should be the deciding factor, I see (1) as the most important one.

Reply:
We would also need to confirm that we can still support sharedcomponent with moving the API to host. The sharedcomponent requirements were not fully understood until sometime after https://github.com/open-telemetry/opentelemetry-collector/commit/b45540e44bd7ad7d114baa983241f7995d72ca5d. The requirements I'm referring to are discussed here: https://github.com/open-telemetry/opentelemetry-collector/pull/10059#issuecomment-2091896948. They are also outlined in the new document: https://github.com/open-telemetry/opentelemetry-collector/blob/main/docs/component-status.md#implementation-details.

Reply:
My first impression is that supporting sharedcomponent will be a challenge if we move ReportStatus from TelemetrySettings to Host. We were able to solve this problem from within sharedcomponent by chaining together the ReportStatus functions for each instance the component represents. This is done through the TelemetrySettings instances passed to LoadOrStore which is called during create. It's important that we have a ReportStatus function that can report for all instances of a sharecomponent before start-time, because the component needs to report the starting events for its instances at the same time. Trying to discover the instances from within the sharedcomponent at start-time will not work because its already too late. We would need a way to know what instances a sharedcomponent represents from "outside" the sharedcomponent to make this work. That said, the current sharedcomponent situation is not ideal and perhaps needs improvement anyways.

Reply:
It also looks like the status reporting parts of shared component only made it into Core's internal SharedComponent, not contrib, so only the otlpreceiver is impacted by SharedComponent.

Reply:
I've made more progress on the sharedcomponent problem.  It isn't perfect, but I think it is acceptable.  I hope to have a PR up by the end of the week for us to discuss, and then if it is acceptable I believe there are some incremental PRs possible to actually introduce the changes.

Reply:
Here is an attempt: https://github.com/open-telemetry/opentelemetry-collector/pull/10725.

Reply:
The initial feedback in https://github.com/open-telemetry/opentelemetry-collector/pull/10725 is promising, so I'm going to call this conversation resolved